## 4. Advanced Association Rule Minning

### a) 

In [1]:
import pandas as pd
import numpy as np
import jenkspy


In [2]:
rdf = pd.read_csv('Heart.csv')
rdf.head(3)

age  gender  chest pain type  serum cholestoral   fasting blood sugar  \
0   63       1                3                 233                    1   
1   37       1                2                 250                    0   
2   41       0                1                 204                    0   

   resting electrocardiographic  maximum heart rate achieved  \
0                             0                          NaN   
1                             1                        187.0   
2                             0                        172.0   

   exercise induced angina  target  
0                        0       1  
1                        0       1  
2                        0       1

In [3]:
rdf.corr()

age    gender  chest pain type  \
age                           1.000000 -0.098447        -0.068653   
gender                       -0.098447  1.000000        -0.049353   
chest pain type              -0.068653 -0.049353         1.000000   
serum cholestoral             0.213678 -0.197912        -0.076904   
fasting blood sugar           0.121308  0.045032         0.094444   
resting electrocardiographic -0.116211 -0.058196         0.044421   
maximum heart rate achieved  -0.399170 -0.044089         0.297586   
exercise induced angina       0.096801  0.141664        -0.394280   
target                       -0.225439 -0.280937         0.433798   

                              serum cholestoral   fasting blood sugar  \
age                                     0.213678             0.121308   
gender                                 -0.197912             0.045032   
chest pain type                        -0.076904             0.094444   
serum cholestoral                       1.000000             0.013294   
fasting blood sugar                     0.013294             1.000000   
resting electrocardiographic           -0.151040            -0.084189   
maximum heart rate achieved            -0.009928            -0.008773   
exercise induced angina                 0.067023             0.025665   
target                                 -0.085239            -0.028046   

                              resting electrocardiographic  \
age                                              -0.116211   
gender                                           -0.058196   
chest pain type                                   0.044421   
serum cholestoral                                -0.151040   
fasting blood sugar                              -0.084189   
resting electrocardiographic                      1.000000   
maximum heart rate achieved                       0.044249   
exercise induced angina                          -0.070733   
target                                            0.137230   

                              maximum heart rate achieved  \
age                                             -0.399170   
gender                                          -0.044089   
chest pain type                                  0.297586   
serum cholestoral                               -0.009928   
fasting blood sugar                             -0.008773   
resting electrocardiographic                     0.044249   
maximum heart rate achieved                      1.000000   
exercise induced angina                         -0.379081   
target                                           0.422279   

                              exercise induced angina    target  
age                                          0.096801 -0.225439  
gender                                       0.141664 -0.280937  
chest pain type                             -0.394280  0.433798  
serum cholestoral                            0.067023 -0.085239  
fasting blood sugar                          0.025665 -0.028046  
resting electrocardiographic                -0.070733  0.137230  
maximum heart rate achieved                 -0.379081  0.422279  
exercise induced angina                      1.000000 -0.436757  
target                                      -0.436757  1.000000

In [4]:
rdf.shape

(303, 9)

In [5]:
rdf.nunique()

age                              41
gender                            2
chest pain type                   4
serum cholestoral               152
fasting blood sugar               2
resting electrocardiographic      3
maximum heart rate achieved      91
exercise induced angina           2
target                            2
dtype: int64

#### Replacing Null Values with Mean:  

In [6]:
rdf.isnull().sum()

age                             0
gender                          0
chest pain type                 0
serum cholestoral               0
fasting blood sugar             0
resting electrocardiographic    0
maximum heart rate achieved     1
exercise induced angina         0
target                          0
dtype: int64

In [7]:
rdf['maximum heart rate achieved'].replace(np.NaN,rdf['maximum heart rate achieved'].mean(),inplace=True)

In [8]:
rdf.isnull().sum()

age                             0
gender                          0
chest pain type                 0
serum cholestoral               0
fasting blood sugar             0
resting electrocardiographic    0
maximum heart rate achieved     0
exercise induced angina         0
target                          0
dtype: int64

#### Binning :

In [9]:
breaks = jenkspy.jenks_breaks(rdf['age'], nb_class=3)
print(breaks)

[29.0, 49.0, 60.0, 77.0]


In [10]:
rdf['age_c']=pd.cut(rdf['age'], bins=breaks,labels=['29<age<49','49<age<60','60<age<77'])

In [11]:
breaks = jenkspy.jenks_breaks(rdf['serum cholestoral '], nb_class=3)
print(breaks)

[126.0, 226.0, 290.0, 564.0]


In [12]:
rdf['serum cholestoral _c']=pd.qcut(rdf['serum cholestoral '], q=3,labels=['126<sc<226','226<sc<290','290<sc<564'])

In [13]:
breaks = jenkspy.jenks_breaks(rdf['maximum heart rate achieved'], nb_class=3)
print(breaks)

[71.0, 133.0, 160.0, 202.0]


In [14]:
rdf['maximum heart rate achieved_c']=pd.qcut(rdf['maximum heart rate achieved'], q=3,labels=['71<max<133','133<max<160','160<max<202'])

In [15]:
rdf.drop(['age','serum cholestoral ','maximum heart rate achieved'],axis=1,inplace=True)
rdf

gender  chest pain type  fasting blood sugar  \
0         1                3                    1   
1         1                2                    0   
2         0                1                    0   
3         1                1                    0   
4         0                0                    0   
..      ...              ...                  ...   
298       0                0                    0   
299       1                3                    0   
300       1                0                    1   
301       1                0                    0   
302       0                1                    0   

     resting electrocardiographic  exercise induced angina  target      age_c  \
0                               0                        0       1  60<age<77   
1                               1                        0       1  29<age<49   
2                               0                        0       1  29<age<49   
3                               1                        0       1  49<age<60   
4                               1                        1       1  49<age<60   
..                            ...                      ...     ...        ...   
298                             1                        1       0  49<age<60   
299                             1                        0       0  29<age<49   
300                             1                        0       0  60<age<77   
301                             1                        1       0  49<age<60   
302                             0                        0       0  49<age<60   

    serum cholestoral _c maximum heart rate achieved_c  
0             226<sc<290                   133<max<160  
1             226<sc<290                   160<max<202  
2             126<sc<226                   160<max<202  
3             226<sc<290                   160<max<202  
4             290<sc<564                   160<max<202  
..                   ...                           ...  
298           226<sc<290                    71<max<133  
299           290<sc<564                    71<max<133  
300           126<sc<226                    71<max<133  
301           126<sc<226                    71<max<133  
302           226<sc<290                   160<max<202  

[303 rows x 9 columns]

#### Binarizing Dataframe:

In [16]:
rdf.nunique()

gender                           2
chest pain type                  4
fasting blood sugar              2
resting electrocardiographic     3
exercise induced angina          2
target                           2
age_c                            3
serum cholestoral _c             3
maximum heart rate achieved_c    3
dtype: int64

In [17]:
rdf = pd.get_dummies(rdf,columns=['gender','fasting blood sugar','exercise induced angina','chest pain type','resting electrocardiographic','age_c','serum cholestoral _c','maximum heart rate achieved_c','target'])
rdf

gender_0  gender_1  fasting blood sugar_0  fasting blood sugar_1  \
0           0         1                      0                      1   
1           0         1                      1                      0   
2           1         0                      1                      0   
3           0         1                      1                      0   
4           1         0                      1                      0   
..        ...       ...                    ...                    ...   
298         1         0                      1                      0   
299         0         1                      1                      0   
300         0         1                      0                      1   
301         0         1                      1                      0   
302         1         0                      1                      0   

     exercise induced angina_0  exercise induced angina_1  chest pain type_0  \
0                            1                          0                  0   
1                            1                          0                  0   
2                            1                          0                  0   
3                            1                          0                  0   
4                            0                          1                  1   
..                         ...                        ...                ...   
298                          0                          1                  1   
299                          1                          0                  0   
300                          1                          0                  1   
301                          0                          1                  1   
302                          1                          0                  0   

     chest pain type_1  chest pain type_2  chest pain type_3  ...  \
0                    0                  0                  1  ...   
1                    0                  1                  0  ...   
2                    1                  0                  0  ...   
3                    1                  0                  0  ...   
4                    0                  0                  0  ...   
..                 ...                ...                ...  ...   
298                  0                  0                  0  ...   
299                  0                  0                  1  ...   
300                  0                  0                  0  ...   
301                  0                  0                  0  ...   
302                  1                  0                  0  ...   

     age_c_49<age<60  age_c_60<age<77  serum cholestoral _c_126<sc<226  \
0                  0                1                                0   
1                  0                0                                0   
2                  0                0                                1   
3                  1                0                                0   
4                  1                0                                0   
..               ...              ...                              ...   
298                1                0                                0   
299                0                0                                0   
300                0                1                                1   
301                1                0                                1   
302                1                0                                0   

     serum cholestoral _c_226<sc<290  serum cholestoral _c_290<sc<564  \
0                                  1                                0   
1                                  1                                0   
2                                  0                                0   
3                                  1                                0   
4                                  0                                1   
..         

### b) 

In [18]:
from mlxtend.frequent_patterns import apriori

In [19]:
frequent_itemsets = apriori(rdf, min_support=0.2, use_colnames=True)
frequent_itemsets

support                                           itemsets
0    0.316832                                         (gender_0)
1    0.683168                                         (gender_1)
2    0.851485                            (fasting blood sugar_0)
3    0.673267                        (exercise induced angina_0)
4    0.326733                        (exercise induced angina_1)
..        ...                                                ...
130  0.201320  (target_0, resting electrocardiographic_0, che...
131  0.207921  (target_0, maximum heart rate achieved_c_71<ma...
132  0.204620  (fasting blood sugar_0, exercise induced angin...
133  0.234323  (target_0, fasting blood sugar_0, gender_1, ch...
134  0.244224  (fasting blood sugar_0, resting electrocardiog...

[135 rows x 2 columns]

### c) 

In [20]:
from mlxtend.frequent_patterns import association_rules

In [21]:
rules = association_rules(frequent_itemsets, metric="lift", min_threshold=1)
rules

antecedents  \
0                  (fasting blood sugar_0)   
1                               (gender_0)   
2              (exercise induced angina_0)   
3                               (gender_0)   
4                               (target_1)   
..                                     ...   
333  (exercise induced angina_0, target_1)   
334                (fasting blood sugar_0)   
335       (resting electrocardiographic_1)   
336            (exercise induced angina_0)   
337                             (target_1)   

                                           consequents  antecedent support  \
0                                           (gender_0)            0.851485   
1                              (fasting blood sugar_0)            0.316832   
2                                           (gender_0)            0.673267   
3                          (exercise induced angina_0)            0.316832   
4                                           (gender_0)            0.544554   
..                                                 ...                 ...   
333  (fasting blood sugar_0, resting electrocardiog...            0.468647   
334  (resting electrocardiographic_1, exercise indu...            0.851485   
335  (fasting blood sugar_0, exercise induced angin...            0.501650   
336  (fasting blood sugar_0, resting electrocardiog...            0.673267   
337  (fasting blood sugar_0, resting electrocardiog...            0.544554   

     consequent support   support  confidence      lift  leverage  conviction  
0              0.316832  0.277228    0.325581  1.027616  0.007450    1.012974  
1              0.851485  0.277228    0.875000  1.027616  0.007450    1.188119  
2              0.316832  0.244224    0.362745  1.144914  0.030912    1.072049  
3              0.673267  0.244224    0.770833  1.144914  0.030912    1.425743  
4              0.316832  0.237624    0.436364  1.377273  0.065092    1.212073  
..                  ...       ...         ...       ...       ...         ...  
333            0.438944  0.244224    0.521127  1.187229  0.038515    1.171617  
334            0.280528  0.244224    0.286822  1.022435  0.005359    1.008825  
335            0.399340  0.244224    0.486842  1.219117  0.043895    1.170517  
336            0.277228  0.244224    0.362745  1.308473  0.057576    1.134196  
337            0.310231  0.244224    0.448485  1.445648  0.075287    1.250680  

[338 rows x 9 columns]

####  Stroke Doesn't Happen:

In [22]:
display(rules[rules["consequents"].apply(lambda x: 'target_0' in str(x))])

antecedents  \
23                                          (gender_1)   
57                         (exercise induced angina_1)   
65                                 (chest pain type_0)   
71                    (resting electrocardiographic_0)   
79                                   (age_c_49<age<60)   
83          (maximum heart rate achieved_c_71<max<133)   
140                  (fasting blood sugar_0, gender_1)   
142                            (fasting blood sugar_0)   
143                                         (gender_1)   
158              (gender_1, exercise induced angina_1)   
160                                         (gender_1)   
161                        (exercise induced angina_1)   
164                      (gender_1, chest pain type_0)   
166                                         (gender_1)   
167                                (chest pain type_0)   
170         (resting electrocardiographic_0, gender_1)   
172                   (resting electrocardiographic_0)   
173                                         (gender_1)   
176  (maximum heart rate achieved_c_71<max<133, gen...   
178         (maximum heart rate achieved_c_71<max<133)   
179                                         (gender_1)   
225  (fasting blood sugar_0, exercise induced angin...   
227                        (exercise induced angina_1)   
241         (fasting blood sugar_0, chest pain type_0)   
243                                (chest pain type_0)   
245  (fasting blood sugar_0, resting electrocardiog...   
247                   (resting electrocardiographic_0)   
254  (maximum heart rate achieved_c_71<max<133, fas...   
257         (maximum heart rate achieved_c_71<max<133)   
258                            (fasting blood sugar_0)   
286     (chest pain type_0, exercise induced angina_1)   
288                                (chest pain type_0)   
289                        (exercise induced angina_1)   
292  (resting electrocardiographic_0, chest pain ty...   
294                   (resting electrocardiographic_0)   
295                                (chest pain type_0)   
298  (maximum heart rate achieved_c_71<max<133, che...   
300         (maximum heart rate achieved_c_71<max<133)   
301                                (chest pain type_0)   
313  (fasting blood sugar_0, gender_1, chest pain t...   
317                  (fasting blood sugar_0, gender_1)   
318         (fasting blood sugar_0, chest pain type_0)   
319                      (gender_1, chest pain type_0)   
321                            (fasting blood sugar_0)   
322                                         (gender_1)   
323                                (chest pain type_0)   

                                           consequents  antecedent support  \
23                                          (target_0)            0.683168   
57                                          (target_0)            0.326733   
65                                          (target_0)            0.471947   
71                                          (target_0)            0.485149   
79                                          (target_0)            0.448845   
83                                          (target_0)            0.353135   
140                                         (target_0)            0.574257   
142                               (target_0, gender_1)            0.851485   
143                  (target_0, fasting blood sugar_0)            0.683168   
158                                         (target_0)            0.254125   
160              (target_0, exercise induced angina_1)            0.683168   
161                               (target_0, gender_1)            0.326733   
164                                         (target_0)            0.343234   
166                      (target_0, chest pain type_0)            0.683168   
167                               (target_0, gender_1)            0.471947   
170                                         (target_0)            0.

#### Stroke Happens: 

In [23]:
res=rules[rules["consequents"].apply(lambda x: 'target_1' in str(x))]
display(res)

antecedents  \
5                                           (gender_0)   
36                             (fasting blood sugar_0)   
52                         (exercise induced angina_0)   
66                                 (chest pain type_2)   
74                    (resting electrocardiographic_1)   
77                                   (age_c_29<age<49)   
80                   (serum cholestoral _c_126<sc<226)   
84         (maximum heart rate achieved_c_160<max<202)   
93                   (fasting blood sugar_0, gender_0)   
95                             (fasting blood sugar_0)   
97                                          (gender_0)   
99               (exercise induced angina_0, gender_0)   
101                        (exercise induced angina_0)   
103                                         (gender_0)   
146              (exercise induced angina_0, gender_1)   
148                        (exercise induced angina_0)   
214  (fasting blood sugar_0, exercise induced angin...   
217                            (fasting blood sugar_0)   
218                        (exercise induced angina_0)   
248  (fasting blood sugar_0, resting electrocardiog...   
251                            (fasting blood sugar_0)   
252                   (resting electrocardiographic_1)   
260  (maximum heart rate achieved_c_160<max<202, fa...   
263        (maximum heart rate achieved_c_160<max<202)   
264                            (fasting blood sugar_0)   
266     (chest pain type_2, exercise induced angina_0)   
269                                (chest pain type_2)   
270                        (exercise induced angina_0)   
272  (resting electrocardiographic_1, exercise indu...   
275                   (resting electrocardiographic_1)   
276                        (exercise induced angina_0)   
278  (maximum heart rate achieved_c_160<max<202, ex...   
281        (maximum heart rate achieved_c_160<max<202)   
282                        (exercise induced angina_0)   
302  (fasting blood sugar_0, exercise induced angin...   
304  (fasting blood sugar_0, exercise induced angin...   
306              (exercise induced angina_0, gender_1)   
308                        (exercise induced angina_0)   
324  (fasting blood sugar_0, resting electrocardiog...   
328  (fasting blood sugar_0, resting electrocardiog...   
329  (fasting blood sugar_0, exercise induced angin...   
331  (resting electrocardiographic_1, exercise indu...   
334                            (fasting blood sugar_0)   
335                   (resting electrocardiographic_1)   
336                        (exercise induced angina_0)   

                                           consequents  antecedent support  \
5                                           (target_1)            0.316832   
36                                          (target_1)            0.851485   
52                                          (target_1)            0.673267   
66                                          (target_1)            0.287129   
74                                          (target_1)            0.501650   
77                                          (target_1)            0.287129   
80                                          (target_1)            0.333333   
84                                          (target_1)            0.310231   
93                                          (target_1)            0.277228   
95                                (gender_0, target_1)            0.851485   
97                   (fasting blood sugar_0, target_1)            0.316832   
99                                          (target_1)            0.244224   
101                               (gender_0, target_1)            0.673267   
103              (exercise induced angina_0, target_1)            0.316832   
146                                         (target_1)            0.429043   
148                               (gender_1, target_1)            0.673267   
214                                         (tar

### d) 

<hr style = "border-top: 3px solid #000000 ; border-radius: 3px;">

<p style =" direction:rtl;text-align:right;">

    84	(maximum heart rate achieved_c_160<max<202)	(target_1)	0.310231	0.544554	0.244224	0.787234	1.445648	0.075287	2.140594
</p>


<p style =" direction:rtl;text-align:right;">
 در مثال فوق اگر بیشینه heart rate بین 160 و 202  باشد باعث وقوع سکته میشود.
</p>

<p style =" direction:rtl;text-align:right;">
این سطر بیانگر این است که  این بازه heart rate به طور مکرر رخ داده است
ساپورت برای آن  برابر 0.310231 است که فرکانس نسبی تکرار این رابطه (کاندیدا) است.
این عدد برای target_1 هم مجزا در قابل consequent support گزارش شده است.
support به تنهایی هردو این محموعه ها را در نظر میگیرد و برای اجتماع آنها این عدد را 0.244224 گزارش می کند.
کانفیدنس برای این قاعده 0.787234 است یعنی بین همه افرادی که  این بازه ضزبان را تحربه کرده اند 78درصد آنها به سکته قلبی مبتلا شده اند.
    
</p>






<p style =" direction:rtl;text-align:right;"> از متریک لیفت معمولاً برای اندازه گیری اینکه antecedents و در consequents یک قانون A-> C اغلب چقدر باهم اتفاق می افتد استفاده می شود ، بیش از آنکه انتظار داشته باشیم اگر از نظر آماری مستقل باشند. 
    اگر A و C مستقل باشند ، امتیاز Lift دقیقاً 1 خواهد بود.
</p>

<p style =" direction:rtl;text-align:right;">leverage تفاوت بین فرکانس مشاهده شده A و C با هم و فرکانسی را که در صورت استقلال A و C انتظار می رود محاسبه می کند. مقدار  0 نشانگر استقلال است.

</p>

<p style =" direction:rtl;text-align:right;">مقدار conviction زیاد به این معنی است که consequents بسیار بستگی به antecedents دارد. به عنوان مثال ، در مورد confidence کامل ، مخرج 0 می شود (به دلیل 1 - 1) که  conviction برای آن "inf" تعریف شده است. مشابه lift ، اگر اقلام مستقل باشند ، محکومیت 1 است.
</p>
<hr style = "border-top: 3px solid #000000 ; border-radius: 3px;">
